File: ruby/test/registration_test.rb
# frozen_string_literal: true

require_relative "test_helper"
require "tmpdir"
require "fileutils"
require "openssl"

class RegistrationTest < Minitest::Test
  SUBSTRATE_ENDPOINT = "https://app.datagrout.ai/api/v1/substrate/identity"
  CA_URL = "https://ca.datagrout.ai/ca.pem"
  SERVER_URL = "https://gateway.datagrout.ai/servers/test-uuid/mcp"
  TOKEN_ENDPOINT = "https://gateway.datagrout.ai/servers/test-uuid/oauth/token"

  # Fake PEMs for tests that only check string content, not OpenSSL parsing.
  SAMPLE_CERT = <<~PEM
    -----BEGIN CERTIFICATE-----
    MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA
    -----END CERTIFICATE-----
  PEM

  SAMPLE_KEY = <<~PEM
    -----BEGIN EC PRIVATE KEY-----
    MIIEvAIBADANBgkqhkiG9w0BAQEFAASCBKYwggSiAgEA
    -----END EC PRIVATE KEY-----
  PEM

  SAMPLE_CA = <<~PEM
    -----BEGIN CERTIFICATE-----
    MIIBkDCB+gIJALRiMLAh1kkKMA0GCSqGSIb3DQEBCwUA
    -----END CERTIFICATE-----
  PEM

  # Generate a valid self-signed cert + key for tests that build a full Client
  # (which invokes OpenSSL parsing through the transport layer).
  def self.generate_real_identity
    key = OpenSSL::PKey::EC.generate("prime256v1")
    cert = OpenSSL::X509::Certificate.new
    cert.version = 2
    cert.serial = 1
    cert.subject = OpenSSL::X509::Name.parse("/CN=test-conduit")
    cert.issuer = cert.subject
    cert.not_before = Time.now - 3600
    cert.not_after = Time.now + 86_400 * 365
    cert.public_key = key
    cert.sign(key, OpenSSL::Digest::SHA256.new)
    [cert.to_pem, key.to_pem]
  end

  REAL_CERT, REAL_KEY = generate_real_identity

  def setup
    WebMock.disable_net_connect!
  end

  def teardown
    WebMock.reset!
    WebMock.allow_net_connect!
  end

  # ================================================================
  # generate_keypair
  # ================================================================

  def test_generate_keypair_returns_two_pem_strings
    private_pem, public_pem = DatagroutConduit::Registration.generate_keypair

    assert private_pem.include?("-----BEGIN EC PRIVATE KEY-----") ||
           private_pem.include?("-----BEGIN PRIVATE KEY-----"),
           "Expected private key PEM header"

    assert public_pem.include?("-----BEGIN PUBLIC KEY-----") ||
           public_pem.include?("-----BEGIN EC PRIVATE KEY-----"),
           "Expected public key PEM header"
  end

  def test_generate_keypair_produces_valid_ec_key
    private_pem, _public_pem = DatagroutConduit::Registration.generate_keypair

    key = OpenSSL::PKey.read(private_pem)
    assert_kind_of OpenSSL::PKey::EC, key
    assert_equal "prime256v1", key.group.curve_name
  end

  def test_generate_keypair_produces_unique_keys
    pair_a = DatagroutConduit::Registration.generate_keypair
    pair_b = DatagroutConduit::Registration.generate_keypair

    refute_equal pair_a[0], pair_b[0], "Two generated keypairs should differ"
  end

  # ================================================================
  # save_identity
  # ================================================================

  def test_save_identity_creates_files
    Dir.mktmpdir do |dir|
      paths = DatagroutConduit::Registration.save_identity(SAMPLE_CERT, SAMPLE_KEY, dir)

      assert File.exist?(paths[:cert])
      assert File.exist?(paths[:key])
      assert_equal SAMPLE_CERT, File.read(paths[:cert])
      assert_equal SAMPLE_KEY, File.read(paths[:key])
      assert_nil paths[:ca]
    end
  end

  def test_save_identity_creates_ca_file
    Dir.mktmpdir do |dir|
      paths = DatagroutConduit::Registration.save_identity(SAMPLE_CERT, SAMPLE_KEY, dir, ca_pem: SAMPLE_CA)

      assert File.exist?(paths[:ca])
      assert_equal SAMPLE_CA, File.read(paths[:ca])
    end
  end

  def test_save_identity_sets_permissions
    Dir.mktmpdir do |dir|
      paths = DatagroutConduit::Registration.save_identity(SAMPLE_CERT, SAMPLE_KEY, dir, ca_pem: SAMPLE_CA)

      assert_equal 0o600, File.stat(paths[:cert]).mode & 0o777
      assert_equal 0o600, File.stat(paths[:key]).mode & 0o777
      assert_equal 0o600, File.stat(paths[:ca]).mode & 0o777
    end
  end

  def test_save_identity_creates_dir_if_missing
    Dir.mktmpdir do |base|
      nested = File.join(base, "a", "b", "c")
      paths = DatagroutConduit::Registration.save_identity(SAMPLE_CERT, SAMPLE_KEY, nested)

      assert File.directory?(nested)
      assert File.exist?(paths[:cert])
    end
  end

  def test_save_identity_uses_standard_filenames
    Dir.mktmpdir do |dir|
      paths = DatagroutConduit::Registration.save_identity(SAMPLE_CERT, SAMPLE_KEY, dir, ca_pem: SAMPLE_CA)

      assert_equal File.join(dir, "identity.pem"), paths[:cert]
      assert_equal File.join(dir, "identity_key.pem"), paths[:key]
      assert_equal File.join(dir, "ca.pem"), paths[:ca]
    end
  end

  # ================================================================
  # fetch_ca_cert
  # ================================================================

  def test_fetch_ca_cert_success
    stub_request(:get, CA_URL)
      .to_return(
        status: 200,
        body: SAMPLE_CA
      )

    pem = DatagroutConduit::Registration.fetch_ca_cert
    assert_equal SAMPLE_CA, pem
  end

  def test_fetch_ca_cert_custom_url
    custom_url = "https://custom-ca.example.com/ca.pem"
    stub_request(:get, custom_url)
      .to_return(status: 200, body: SAMPLE_CA)

    pem = DatagroutConduit::Registration.fetch_ca_cert(ca_url: custom_url)
    assert_equal SAMPLE_CA, pem
  end

  def test_fetch_ca_cert_raises_on_failure
    stub_request(:get, CA_URL).to_return(status: 500, body: "error")

    assert_raises(DatagroutConduit::ConnectionError) do
      DatagroutConduit::Registration.fetch_ca_cert
    end
  end

  def test_fetch_ca_cert_raises_on_invalid_pem
    stub_request(:get, CA_URL).to_return(status: 200, body: "not a cert")

    assert_raises(DatagroutConduit::ConnectionError) do
      DatagroutConduit::Registration.fetch_ca_cert
    end
  end

  # ================================================================
  # refresh_ca_cert
  # ================================================================

  def test_refresh_ca_cert_writes_file
    stub_request(:get, CA_URL).to_return(status: 200, body: SAMPLE_CA)

    Dir.mktmpdir do |dir|
      path = DatagroutConduit::Registration.refresh_ca_cert(dir)

      assert_equal File.join(dir, "ca.pem"), path
      assert_equal SAMPLE_CA, File.read(path)
      assert_equal 0o600, File.stat(path).mode & 0o777
    end
  end

  # ================================================================
  # register_identity
  # ================================================================

  def test_register_identity_success
    stub_request(:post, "#{SUBSTRATE_ENDPOINT}/register")
      .with(
        headers: { "Authorization" => "Bearer test-token" }
      )
      .to_return(
        status: 201,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(
          "id" => "sub_123",
          "cert_pem" => SAMPLE_CERT,
          "ca_cert_pem" => SAMPLE_CA,
          "fingerprint" => "abc123",
          "name" => "conduit-client",
          "registered_at" => "2026-03-01T00:00:00Z",
          "valid_until" => "2026-03-31T00:00:00Z"
        )
      )

    _private_pem, public_pem = DatagroutConduit::Registration.generate_keypair
    reg = DatagroutConduit::Registration.register_identity(
      public_pem,
      auth_token: "test-token"
    )

    assert_equal "sub_123", reg.id
    assert_equal SAMPLE_CERT, reg.cert_pem
    assert_equal SAMPLE_CA, reg.ca_cert_pem
    assert_equal "abc123", reg.fingerprint
    assert_equal "conduit-client", reg.name
    assert_equal "2026-03-31T00:00:00Z", reg.valid_until
  end

  def test_register_identity_failure
    stub_request(:post, "#{SUBSTRATE_ENDPOINT}/register")
      .to_return(status: 401, body: "Unauthorized")

    assert_raises(DatagroutConduit::AuthError) do
      DatagroutConduit::Registration.register_identity(
        "PUBLIC KEY PEM",
        auth_token: "bad-token"
      )
    end
  end

  # ================================================================
  # default_identity_dir
  # ================================================================

  def test_default_identity_dir
    dir = DatagroutConduit::Registration.default_identity_dir
    refute_nil dir
    assert dir.end_with?(".conduit")
  end

  # ================================================================
  # Constants
  # ================================================================

  def test_constants_accessible_from_module
    assert_equal "https://ca.datagrout.ai/ca.pem", DatagroutConduit::DG_CA_URL
    assert_equal "https://app.datagrout.ai/api/v1/substrate/identity",
                 DatagroutConduit::DG_SUBSTRATE_ENDPOINT
  end

  # ================================================================
  # bootstrap_identity
  # ================================================================

  def test_bootstrap_identity_uses_existing_identity
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, "identity.pem"), REAL_CERT)
      File.write(File.join(dir, "identity_key.pem"), REAL_KEY)

      client = DatagroutConduit::Client.bootstrap_identity(
        url: SERVER_URL,
        auth_token: "test-token",
        identity_dir: dir
      )

      assert_kind_of DatagroutConduit::Client, client
    end
  end

  def test_bootstrap_identity_registers_new_identity
    Dir.mktmpdir do |dir|
      stub_request(:post, "#{SUBSTRATE_ENDPOINT}/register")
        .to_return(
          status: 201,
          headers: { "Content-Type" => "application/json" },
          body: JSON.generate(
            "id" => "sub_new",
            "cert_pem" => REAL_CERT,
            "ca_cert_pem" => nil,
            "fingerprint" => "xyz789",
            "name" => "conduit-client",
            "registered_at" => "2026-03-01T00:00:00Z",
            "valid_until" => "2026-03-31T00:00:00Z"
          )
        )

      client = DatagroutConduit::Client.bootstrap_identity(
        url: SERVER_URL,
        auth_token: "test-token",
        name: "test-client",
        identity_dir: dir
      )

      assert_kind_of DatagroutConduit::Client, client
      assert File.exist?(File.join(dir, "identity.pem"))
      assert File.exist?(File.join(dir, "identity_key.pem"))
    end
  end

  # ================================================================
  # bootstrap_identity_oauth
  # ================================================================

  def test_bootstrap_identity_oauth_flow
    Dir.mktmpdir do |dir|
      stub_request(:post, TOKEN_ENDPOINT)
        .to_return(
          status: 200,
          headers: { "Content-Type" => "application/json" },
          body: JSON.generate(
            "access_token" => "oauth_jwt_token",
            "token_type" => "bearer",
            "expires_in" => 3600
          )
        )

      stub_request(:post, "#{SUBSTRATE_ENDPOINT}/register")
        .with(headers: { "Authorization" => "Bearer oauth_jwt_token" })
        .to_return(
          status: 201,
          headers: { "Content-Type" => "application/json" },
          body: JSON.generate(
            "id" => "sub_oauth",
            "cert_pem" => REAL_CERT,
            "ca_cert_pem" => nil,
            "fingerprint" => "oauth_fp",
            "name" => "conduit-client",
            "registered_at" => "2026-03-01T00:00:00Z",
            "valid_until" => "2026-03-31T00:00:00Z"
          )
        )

      client = DatagroutConduit::Client.bootstrap_identity_oauth(
        url: SERVER_URL,
        client_id: "my_id",
        client_secret: "my_secret",
        identity_dir: dir
      )

      assert_kind_of DatagroutConduit::Client, client
      assert File.exist?(File.join(dir, "identity.pem"))
      assert File.exist?(File.join(dir, "identity_key.pem"))
    end
  end

  def test_bootstrap_identity_oauth_propagates_auth_error
    stub_request(:post, TOKEN_ENDPOINT)
      .to_return(
        status: 401,
        headers: { "Content-Type" => "application/json" },
        body: '{"error": "invalid_client"}'
      )

    Dir.mktmpdir do |dir|
      assert_raises(DatagroutConduit::AuthError) do
        DatagroutConduit::Client.bootstrap_identity_oauth(
          url: SERVER_URL,
          client_id: "bad_id",
          client_secret: "bad_secret",
          identity_dir: dir
        )
      end
    end
  end

  # ================================================================
  # RegistrationResponse struct
  # ================================================================

  def test_registration_response_struct
    resp = DatagroutConduit::RegistrationResponse.new(
      id: "sub_1",
      cert_pem: SAMPLE_CERT,
      ca_cert_pem: SAMPLE_CA,
      fingerprint: "fp_123",
      name: "test",
      registered_at: "2026-03-01T00:00:00Z",
      valid_until: "2026-03-31T00:00:00Z"
    )

    assert_equal "sub_1", resp.id
    assert_equal SAMPLE_CERT, resp.cert_pem
    assert_equal SAMPLE_CA, resp.ca_cert_pem
    assert_equal "fp_123", resp.fingerprint
    assert_equal "test", resp.name
    assert_equal "2026-03-01T00:00:00Z", resp.registered_at
    assert_equal "2026-03-31T00:00:00Z", resp.valid_until
  end
end

File: elixir/test/registration_test.exs
defmodule DatagroutConduit.RegistrationTest do
  use ExUnit.Case, async: true

  alias DatagroutConduit.Registration
  alias DatagroutConduit.Registration.RegistrationResponse

  describe "generate_keypair/0" do
    test "returns {:ok, {private_pem, public_pem}}" do
      assert {:ok, {private_pem, public_pem}} = Registration.generate_keypair()
      assert is_binary(private_pem)
      assert is_binary(public_pem)
    end

    test "private key is valid PEM" do
      {:ok, {private_pem, _}} = Registration.generate_keypair()
      assert String.contains?(private_pem, "-----BEGIN EC PRIVATE KEY-----")
      assert String.contains?(private_pem, "-----END EC PRIVATE KEY-----")

      entries = :public_key.pem_decode(private_pem)
      assert length(entries) == 1
      {type, _der, _} = hd(entries)
      assert type == :ECPrivateKey
    end

    test "public key is valid PEM" do
      {:ok, {_, public_pem}} = Registration.generate_keypair()
      assert String.contains?(public_pem, "-----BEGIN PUBLIC KEY-----")
      assert String.contains?(public_pem, "-----END PUBLIC KEY-----")

      entries = :public_key.pem_decode(public_pem)
      assert length(entries) == 1
      {type, _der, _} = hd(entries)
      assert type == :SubjectPublicKeyInfo
    end

    test "generates unique keypairs each time" do
      {:ok, {priv1, pub1}} = Registration.generate_keypair()
      {:ok, {priv2, pub2}} = Registration.generate_keypair()

      refute priv1 == priv2
      refute pub1 == pub2
    end
  end

  describe "save_identity/4" do
    test "writes cert, key, and ca files to directory" do
      dir = temp_dir()

      {:ok, {private_pem, _}} = Registration.generate_keypair()
      cert_pem = "-----BEGIN CERTIFICATE-----\nfake-cert\n-----END CERTIFICATE-----\n"
      ca_pem = "-----BEGIN CERTIFICATE-----\nfake-ca\n-----END CERTIFICATE-----\n"

      assert {:ok, paths} = Registration.save_identity(cert_pem, private_pem, ca_pem, dir)

      assert File.exists?(paths.cert)
      assert File.exists?(paths.key)
      assert File.exists?(paths.ca)

      assert File.read!(paths.cert) == cert_pem
      assert File.read!(paths.key) == private_pem
      assert File.read!(paths.ca) == ca_pem

      File.rm_rf!(dir)
    end

    test "sets key file permissions to 0o600" do
      dir = temp_dir()

      {:ok, {private_pem, _}} = Registration.generate_keypair()
      cert_pem = "-----BEGIN CERTIFICATE-----\nfake\n-----END CERTIFICATE-----\n"

      {:ok, paths} = Registration.save_identity(cert_pem, private_pem, nil, dir)

      {:ok, stat} = File.stat(paths.key)
      assert stat.access == :read_write

      File.rm_rf!(dir)
    end

    test "creates directory if it does not exist" do
      dir = Path.join(System.tmp_dir!(), "conduit_reg_test_nested_#{:rand.uniform(100_000)}/sub/dir")
      refute File.exists?(dir)

      {:ok, {private_pem, _}} = Registration.generate_keypair()
      cert_pem = "-----BEGIN CERTIFICATE-----\nfake\n-----END CERTIFICATE-----\n"

      assert {:ok, _} = Registration.save_identity(cert_pem, private_pem, nil, dir)
      assert File.exists?(dir)

      File.rm_rf!(Path.join(System.tmp_dir!(), "conduit_reg_test_nested_#{:rand.uniform(100_000)}"))
      File.rm_rf!(dir)
    end

    test "handles nil ca_pem" do
      dir = temp_dir()

      {:ok, {private_pem, _}} = Registration.generate_keypair()
      cert_pem = "-----BEGIN CERTIFICATE-----\nfake\n-----END CERTIFICATE-----\n"

      assert {:ok, paths} = Registration.save_identity(cert_pem, private_pem, nil, dir)
      assert paths.ca == nil
      refute File.exists?(Path.join(dir, "ca.pem"))

      File.rm_rf!(dir)
    end
  end

  describe "constants" do
    test "dg_ca_url/0 returns expected URL" do
      assert Registration.dg_ca_url() == "https://ca.datagrout.ai/ca.pem"
    end

    test "dg_substrate_endpoint/0 returns expected URL" do
      assert Registration.dg_substrate_endpoint() == "https://app.datagrout.ai/api/v1/substrate/identity"
    end
  end

  describe "default_identity_dir/0" do
    test "returns a path under home directory" do
      dir = Registration.default_identity_dir()
      assert is_binary(dir)
      assert String.ends_with?(dir, "/.conduit")
    end
  end

  describe "RegistrationResponse struct" do
    test "has all expected fields" do
      resp = %RegistrationResponse{
        id: "sub-123",
        cert_pem: "cert",
        ca_cert_pem: "ca",
        fingerprint: "abc",
        name: "test",
        registered_at: "2026-01-01T00:00:00Z",
        valid_until: "2026-02-01T00:00:00Z"
      }

      assert resp.id == "sub-123"
      assert resp.cert_pem == "cert"
      assert resp.ca_cert_pem == "ca"
      assert resp.fingerprint == "abc"
      assert resp.name == "test"
    end
  end

  defp temp_dir do
    dir = Path.join(System.tmp_dir!(), "conduit_reg_test_#{:rand.uniform(100_000)}")
    File.mkdir_p!(dir)
    dir
  end
end
